extensions-core/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/supervisor/KafkaSupervisor.java

```diff
@@ -0,0 +1,1266 @@
+          taskGroup.tasks.put(taskId, new TaskData());
+        }
+        return;
+      }
+    }
+
+    log.info("Creating new pending completion task group for discovered task [%s]", taskId);
+    TaskGroup newTaskGroup = new TaskGroup(startingPartitions);
+    newTaskGroup.tasks.put(taskId, new TaskData());
+    newTaskGroup.completionTimeout = DateTime.now().plus(ioConfig.getCompletionTimeout());
+
+    taskGroupList.add(newTaskGroup);
+  }
+
+  private void updateTaskStatus()
+  {
+    // update status (and startTime if unknown) of current tasks in taskGroups
+    for (TaskGroup group : taskGroups.values()) {
+      for (Map.Entry<String, TaskData> entry : group.tasks.entrySet()) {
+        String taskId = entry.getKey();
+        TaskData taskData = entry.getValue();
+
+        if (taskData.startTime == null) {
+          try {
+            taskData.startTime = getTaskStartTime(taskId);
+            if (taskData.startTime != null) {
+              long millisRemaining = ioConfig.getTaskDuration().getMillis() - (System.currentTimeMillis()
+                                                                               - taskData.startTime.getMillis());
+              if (millisRemaining > 0) {
+                scheduledExec.schedule(buildRunTask(), millisRemaining, TimeUnit.MILLISECONDS);
+              }
+            }
+          }
+          catch (Exception e) {
+            log.warn(e, "Task [%s] failed to return start time, killing task", taskId);
+            killTask(taskId);
+          }
+        }
+
+        taskData.status = taskStorage.getStatus(taskId).get();
+      }
+    }
+
+    // update status of pending completion tasks in pendingCompletionTaskGroups
+    for (List<TaskGroup> taskGroups : pendingCompletionTaskGroups.values()) {
+      for (TaskGroup group : taskGroups) {
+        for (Map.Entry<String, TaskData> entry : group.tasks.entrySet()) {
+          entry.getValue().status = taskStorage.getStatus(entry.getKey()).get();
+        }
+      }
+    }
+  }
+
+  private void checkTaskDuration()
+  {
+    Iterator<Map.Entry<Integer, TaskGroup>> i = taskGroups.entrySet().iterator();
+    while (i.hasNext()) {
+      Map.Entry<Integer, TaskGroup> groupEntry = i.next();
+      Integer groupId = groupEntry.getKey();
+      TaskGroup group = groupEntry.getValue();
+
+      // find the longest running task from this group
+      DateTime earliestTaskStart = DateTime.now();
+      for (TaskData taskData : group.tasks.values()) {
+        if (earliestTaskStart.isAfter(taskData.startTime)) {
+          earliestTaskStart = taskData.startTime;
+        }
+      }
+
+      // if this task has run longer than the configured duration, signal all tasks in the group to persist
+      if (earliestTaskStart.plus(ioConfig.getTaskDuration()).isBeforeNow()) {
+        log.info("Task group [%d] has run for [%s]", groupId, ioConfig.getTaskDuration());
+        Map<Integer, Long> endOffsets = signalTasksToFinish(groupId);
```

Comment:
Or if a supervisor crashes, in general, while a task group is finishing or partially finishing,

- maybe it had asked half of them to persist but then crashed before the others were asked to persist?
- maybe it had asked all of them to persist, and they're pending but haven't finished yet